scripts/launch_workflow_cf/utils.py

```diff
@@ -5,94 +5,168 @@
-        credentials = GoogleCredentials.get_application_default()
-        credentials = credentials.create_scoped(scopes)
-
-    return credentials.get_access_token().access_token
 
+def prepare_and_launch(workspace_namespace, workspace_name, method_namespace, method_name,
+                       secret_path, workflow_parameters, entity_set_name=None):
+  """Launch a Terra workflow programmatically.
+
+    Arguments `workflow_parameters` and `entity_set_name` are used to specify the parameters to the
+    workflow. The Terra UI is used to control *all other aspects* of the workflow configuration, such as:
+    * which version of the method is to be used
+    * whether call caching is enabled
+    * whether to delete intermediate outputs
+    * the type of the root entity
+    * how the columns of the root entity map onto the workflow parameters
+
+    Args:
+      workspace_namespace: The project id of the Terra billing project in which the workspace resides.
+      workspace_name: The name of the workspace in which the workflow resides.
+      method_namespace: The namespace of the workflow method.
+      method_name: The name of the workflow method.
+      secret_path: The 'Resource ID' of the service account key stored in Secret Manager. Or, if
+        testing locally, the filepath to the json key for the service account.
+      workflow_parameters: A dictionary of key/value pairs to merge with the inputs from the workflow configuration,
+        overwriting any duplicate keys.
+      entity_set_name: The name of the entity set to be used for all other workflow parameters. Defaults to
+        the most recently created entity set of the root entity type.
+    Returns:
+      None; the side effect is the execution of a parameter-parallel Terra workflow.
+  """
+  # Get access token and and add to headers for requests.
+  headers = {"Authorization": "bearer " + get_access_token(secret_path=secret_path)}
+
+  # Get the workflow config.
+  workflow_config_response = get_workflow_method_config(
+      workspace_namespace=workspace_namespace,
+      workspace_name=workspace_name,
+      method_namespace=method_namespace,
+      method_name=method_name,
+      headers=headers
+  )
+  check_fapi_response(response=workflow_config_response, success_code=200)
+  print("Current default workflow configuration:\n", json.dumps(workflow_config_response.json(), indent=4))
+
+  # Merge the workflow inputs configured in the Terra UI with the items from 'workflow_parameters'
+  # in-place, overwriting existing keys.
+  workflow_config_json = workflow_config_response.json()
+  workflow_config_json["inputs"].update(workflow_parameters)
+  if not "rootEntityType" in workflow_config_json:
+    raise ValueError("""The current default workflow configuration does not have a root entity type specified.
+      Be sure to use the Terra UI at least once to successfully run the workflow with an entity set.""")
+
+  # Configure the entity set to be used.
```

Comment:
seems like there's a ton of whitespace changes - are your tabs 2 spaces rather than 4?

Reply:
ah Google prefers 2 spaces, while PEP8 recommends 4. we use 4 spaces in all our codebase - would you be okay keeping it at 4 spaces?

Reply:
Yes, I ran a code linter on it with a setting for 2 spaces. No problem, I'll switch it back to 4.

This is what I used:
* https://github.com/broadinstitute/ml4h/blob/master/CONTRIBUTING.md#install-pylint
* https://github.com/broadinstitute/ml4h/blob/master/pylintrc
* with a changed value for `indent-string='    '` so that it was 4 spaces instead of 2

Reply:
lovely, thanks so much!